src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -265,18 +295,13 @@
    (shutdown-on-error* shutdown-reason-promise app-context svc-id f nil))
   ([shutdown-reason-promise app-context svc-id f on-error-fn]
    (try
-     ; These would normally be pre-conditions; however, this function needs
-     ; to never throw an exception - since it is often called as a wrapper
-     ; around everything inside a `future`, it is important that this function
-     ; never throw anything (like an AssertionError from a pre-condition),
+     ; This schema check would normally be handled via schematizing the fn itself;
+     ; however, this function needs to never throw an exception - since it is often
+     ; called as a wrapper around everything inside a `future`, it is important
+     ; that this function ; never throw anything (like a schema validation error),
      ; since it is likely to just get lost in a `future`.  Instead,
      ; invalid arguments will simply cause the shutdown promise to be delivered.
-     (assert (instance? Atom app-context))
-     (assert (map? @app-context))
-     (assert (keyword? svc-id))
-     (assert (contains? @app-context svc-id))
-     (assert (ifn? f))
-     (assert ((some-fn nil? ifn?) on-error-fn))
+     (schema/validate a/TrapperkeeperAppContext @app-context)
```

Comment:
Seems like we would be losing some of the validation that was being done previously - i.e., that the `svc-id`, `f`, and `on-error-fn` arguments are all of the type we want them to be.  Is that a problem?